file/file_prefetch_buffer.cc

```diff
@@ -819,8 +819,20 @@ bool FilePrefetchBuffer::TryReadFromCacheUntracked(
       assert(reader != nullptr);
       assert(max_readahead_size_ >= readahead_size_);
 
+      // We disallow async IO for compaction reads since their
+      // latencies are not user-visible
       if (for_compaction) {
-        s = Prefetch(opts, reader, offset, std::max(n, readahead_size_));
+        assert(num_buffers_ == 1);
+        // The readahead size for compaction reads is different and the
+        // historical reason is not clear at the moment. TODO: revisit whether
+        // we want to just use the "regular" logic and readahead by the full
+        // readahize_size_.
+        uint64_t trimmed_readahead_size = 0;
+        if (n < readahead_size_) {
+          trimmed_readahead_size = readahead_size_ - n;
```

Comment:
> Where in the original code shows that compaction is prefetching only readahead_size_ - n but not readahead_size_?

https://github.com/facebook/rocksdb/blob/main/file/file_prefetch_buffer.cc#L823

```cpp
      if (for_compaction) {
        s = Prefetch(opts, reader, offset, std::max(n, readahead_size_));
```

I think you have it correct. The `std::max(n, readahead_size_)` that you mentioned is the reason that the prefetch is only `readahead_size_ - n`. If `n` < `readahead_size_`, we read the original `n` bytes plus `readahead_size_ - n` of prefetched data

Reply:
@anand1976 I am in favor of unifying the logic so that we treat compaction readaheads the same as non-compaction readaheads (e.g. fetch the requested amount + `readahead_size_` in either case). Do you have any objections / what are your thoughts? I think this would simplify our logic and don't see downsides

Reply:
For future readers, offline we decided to unify the logic and use the same readahead amount for compaction reads